### DISCUSSION TITLE
Change only the basic in the PHP migration from 7.4 to 8.1

### DISCUSSION THREAD

**Discussion by amendiola** [rectorphp/rector#7571] [Q&A]
Hi,

I am trying to migrate some PHP projects from 7.4 to 8.1 version with Rector and it is a fantastic tool, but I have a little problem. 
When I execute rector it modifies a lot of files by different rules, but I don't want apply the most of these changes.
¿Are any way to only apply the necessary changes (the deprecations) in a migration from PHP 7.4 to PHP 8.1 ?

Thank you very much.

**Comment by mspirkov** [comment]
It seems that there is no separate set that includes only rules for fixing deprecation errors.
I would recommend that you select the necessary rules from this list: https://getrector.com/find-rule?activeRectorSetGroup=php

**Reply by Kenneth-Sills** [reply]
I've found myself wanting this in the past as well, since I maintain some _very_ large legacy projects.

If someone wants, they can see if @TomasVotruba would bless a separate set or flag for this and then open an issue / PR.

My changes [here](https://github.com/rectorphp/rector-src/pull/7898/changes) located most of the rules for deprecations (though not strictly all of the rules for _required_ changes). Unfortunately, the only indication for these right now are `providedMinPhpVersion()` returning a `PhpVersionFeature::DEPRECATE_*` constant; otherwise, you need to review the whole rule list.

**Comment by signor-pedro** [comment]
This is especially needed in larger companies/corporations, where deploys are frequently 2-phase workflows:
- allow to deploy on higher version, e.g. `"php": ">=8.4 <=8.5"`
- deploy on 8.5
- observe for some time and when OK, tighten constraint to `php": "8.5.*"`

Only the last step lets you use the new language features.

**The ideal solution**: run rector with a `--bc-only` flag / dedicated BC-only sets.

Workaround: I currently pass the `--dry-run` output to Claude and let Claude decide whether the proposed change fixes BC/deprecations. This is costly and ugly 😰

**Reply by Kenneth-Sills** [reply]
You can kind-of do this now, order is important:

```php
return RectorConfig::configure()
    // ...
    ->withSets([
        LevelSetList::UP_TO_PHP_85,
    ])
    ->withPhpVersion(PhpVersion::PHP_74);
```

Rector will run all the rules in the upgrade set while skipping any whose resolution is not compatible with that explicitly specified PHP version. So push this, do the cutover and monitor, then remove that last line. 

However, as noted in https://github.com/rectorphp/rector/issues/9646, Rector will not run any deprecation rules for your upgrade targets, so you'll have to tank the deprecation notices in your logs during the cutover phase.

To be totally clear, it *should* still run rules for actual breaking changes.

Also, if you're not aware, PhpStan also supports explicit version ranges for feature detection. I'd recommend using Symfony's PHP polyfills at runtime when using it, too. They can get sneakily included by a dev dependency and trick PhpStan into thinking you have those functions available.

**Reply by signor-pedro** [reply]
Thanks for the wonderful write-up, especially the dev-dependency warning.
